### PR TITLE
Use modfile to generate the go.mod in GenerateProgram

### PR DESCRIPTION
**Pull request by Frassle**
Rather than writing a go.mod with string lerping, this uses the "golang.org/x/mod/modfile" to construct and render the go.mod files we generate for programs.

### CODE REVIEW

File: pkg/codegen/go/gen_program.go [hunk 1]
```diff
@@ -511,14 +512,14 @@ func GenerateProjectFiles(project workspace.Project, program *pcl.Program) (map[
...
+	err = gomod.AddModuleStmt(project.Name.String())
+	contract.AssertNoErrorf(err, "could not add module statement to go.mod")
+	err = gomod.AddGoStmt("1.20")
+	contract.AssertNoErrorf(err, "could not add Go statement to go.mod")
 
-require (
-	github.com/pulumi/pulumi/sdk/v3 v3.30.0
-`)
+	err = gomod.AddRequire("github.com/pulumi/pulumi/sdk/v3", "v3.30.0")
+	contract.AssertNoErrorf(err, "could not add require statement to go.mod")
```
Comment (lunaris) [inline pkg/codegen/go/gen_program.go:522]:
```suggestion
	contract.AssertNoErrorf(err, "could not add require statement for github.com/pulumi/pulumi/sdk/v3 to go.mod")
```

File: pkg/codegen/go/gen_program.go [hunk 2]
```diff
@@ -563,7 +564,8 @@ require (
 
 			if info, ok := p.Language["go"]; ok {
 				if info, ok := info.(GoPackageInfo); ok && info.ModulePath != "" {
-					fmt.Fprintf(&gomod, " %s v%s\n", info.ModulePath, p.Version.String())
+					err = gomod.AddRequire(info.ModulePath, p.Version.String())
+					contract.AssertNoErrorf(err, "could not add require statement to go.mod")
```
Comment (lunaris) [inline pkg/codegen/go/gen_program.go:568]:
```suggestion
					contract.AssertNoErrorf(err, "could not add require statement for %s to go.mod", info.ModulePath)
```

File: pkg/codegen/go/gen_program.go [hunk 3]
```diff
@@ -593,13 +595,15 @@ require (
 			version = "v" + p.Version.String()
 		}
 		if packageName != "" {
-			fmt.Fprintf(&gomod, "	%s %s\n", packageName, version)
+			err = gomod.AddRequire(packageName, version)
+			contract.AssertNoErrorf(err, "could not add require statement to go.mod")
```
Comment (lunaris) [inline pkg/codegen/go/gen_program.go:599]:
```suggestion
			contract.AssertNoErrorf(err, "could not add require statement for %s to go.mod", packageName)
```

**Review by lunaris — APPROVED** [review]
Love it!

**Review by julienp — APPROVED** [review]

**Pull request merged**